staging/src/k8s.io/dynamic-resource-allocation/controller/controller.go

```diff
@@ -517,50 +525,87 @@ func (ctrl *controller) getParameters(ctx context.Context, claim *resourcev1alph
 	return
 }
 
-func (ctrl *controller) allocateClaim(ctx context.Context,
-	claim *resourcev1alpha2.ResourceClaim, claimParameters interface{},
-	class *resourcev1alpha2.ResourceClass, classParameters interface{},
-	selectedNode string,
-	selectedUser *resourcev1alpha2.ResourceClaimConsumerReference) error {
+func (ctrl *controller) allocateClaims(ctx context.Context, claims claimAllocations, selectedNode string, selectedUser *resourcev1alpha2.ResourceClaimConsumerReference) error {
 	logger := klog.FromContext(ctx)
 
-	if claim.Status.Allocation != nil {
-		// This can happen when two PodSchedulingContext objects trigger
-		// allocation attempts (first one wins) or when we see the
-		// update of the PodSchedulingContext object.
-		logger.V(5).Info("Claim already allocated, nothing to do")
+	var needAllocation claimAllocations
+	for _, claim := range claims {
+		if claim.Claim.Status.Allocation != nil {
+			// This can happen when two PodSchedulingContext objects trigger
+			// allocation attempts (first one wins) or when we see the
+			// update of the PodSchedulingContext object.
+			logger.V(5).Info("Claim is already allocated, skipping allocation", "claim", claim.PodClaimName)
+			continue
+		}
+		needAllocation = append(needAllocation, claim)
+	}
+
+	if len(needAllocation) == 0 {
+		logger.V(5).Info("No claims need allocation, nothing to do")
 		return nil
 	}
 
-	claim = claim.DeepCopy()
-	if !ctrl.hasFinalizer(claim) {
-		// Set finalizer before doing anything. We continue with the updated claim.
-		logger.V(5).Info("Adding finalizer")
-		claim.Finalizers = append(claim.Finalizers, ctrl.finalizer)
-		var err error
-		claim, err = ctrl.kubeClient.ResourceV1alpha2().ResourceClaims(claim.Namespace).Update(ctx, claim, metav1.UpdateOptions{})
-		if err != nil {
-			return fmt.Errorf("add finalizer: %v", err)
+	for _, claimAllocation := range needAllocation {
+		if !ctrl.hasFinalizer(claimAllocation.Claim) {
+			patchFn := func() error {
+				claim := claimAllocation.Claim.DeepCopy()
+				// Set finalizer before doing anything. We continue with the updated claim.
+				logger.V(5).Info("Adding finalizer", "claim", claim.Name)
+				claim.Finalizers = append(claim.Finalizers, ctrl.finalizer)
+				var err error
+				claim, err = ctrl.kubeClient.ResourceV1alpha2().ResourceClaims(claim.Namespace).Update(ctx, claim, metav1.UpdateOptions{})
+				if err != nil {
+					return fmt.Errorf("add finalizer: %v", err)
+				}
+				ctrl.claimCache.Mutation(claim)
+				claimAllocation.Claim = claim
+				return nil
+			}
+			err := retry.RetryOnConflict(retry.DefaultRetry, patchFn)
```

Comment:
`syncClaim -> allocateClaims` is already handling retries. I don't think we should have retry loops inside retry loops.

This inner retry is done without repeating any of the prior checks for the claim. Suppose the `DeletionTimestamp` got set and that caused the conflict. The code above then will add the finalizer although it shouldn't.

Reply:
Indeed, it was not needed, I had weird issues with e2e and unit tests, so this retry was introduced. We narrowed down problem later to be inconsistent (doesn't have to be) behavior of claims cache. I'll remove them, last push was solely to see if the same tests will fail - and now we know that it was not failing for me because I had a single node cluster.
Fixing is on the way.